buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/incubation/IncubatingApiReportTask.kt

```diff
@@ -225,3 +206,169 @@ class Analyzer @Inject constructor(
     private
     fun String.escape() = replace("<", "&lt;").replace(">", "&gt;")
 }
+
+
+private
+interface VersionsToIncubatingCollector {
+
+    fun collectFrom(sourceFile: File): VersionsToIncubating
+}
+
+
+private
+class CompositeVersionsToIncubatingCollector(
+
+    private
+    val collectors: List<VersionsToIncubatingCollector>
+
+) : VersionsToIncubatingCollector {
+
+    override fun collectFrom(sourceFile: File): VersionsToIncubating =
+        collectors
+            .flatMap { it.collectFrom(sourceFile).entries }
+            .associate { it.key to it.value }
+}
+
+
+private
+const val versionNotFound = "Not found"
+
+
+private
+class JavaVersionsToIncubatingCollector(srcDir: File) : VersionsToIncubatingCollector {
+
+    private
+    val solver = JavaSymbolSolver(CombinedTypeSolver(JavaParserTypeSolver(srcDir), ReflectionTypeSolver()))
+
+    override fun collectFrom(sourceFile: File): VersionsToIncubating {
+        val versionsToIncubating = mutableMapOf<Version, MutableSet<IncubatingDescription>>()
+        if (!sourceFile.name.endsWith(".java")) {
+            return versionsToIncubating
+        }
+        JavaParser.parse(sourceFile).run {
+            solver.inject(this)
+            findAllIncubating()
+                .map { node -> toVersionIncubating(sourceFile, node) }
+                .forEach { (version, incubating) ->
+                    versionsToIncubating.getOrPut(version) { mutableSetOf() }.add(incubating)
+                }
+        }
+        return versionsToIncubating
+    }
+
+    private
+    fun CompilationUnit.findAllIncubating() =
+        findAll(Node::class.java).filter { it.isIncubating }
+
+    private
+    val Node.isIncubating: Boolean
+        get() = (this as? NodeWithAnnotations<*>)?.annotations?.any { it.nameAsString == "Incubating" } ?: false
+
+    private
+    fun CompilationUnit.toVersionIncubating(sourceFile: File, node: Node) =
+        Pair(
+            findVersionFromJavadoc(node as NodeWithAnnotations<*>) ?: versionNotFound,
+            nodeName(node, this, sourceFile)
+        )
+
+    private
+    fun findVersionFromJavadoc(node: NodeWithAnnotations<*>): String? = if (node is NodeWithJavadoc<*>) {
```

Comment:
⭕️ Consider replacing `if (e is T) thenBlock else null` by `(e as? T)?.let thenBlock`